core/app/components/Memberships/MemberInviteForm.tsx

```diff
@@ -27,48 +27,69 @@ import { Input } from "ui/input";
 import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "ui/select";
 import { toast } from "ui/use-toast";
 
-import type { MemberFormState } from "./AddMember";
 import { useCommunity } from "~/app/components/providers/CommunityProvider";
+import { client } from "~/lib/api";
 import { didSucceed, useServerAction } from "~/lib/serverActions";
-import * as actions from "./actions";
 import { memberInviteFormSchema } from "./memberInviteFormSchema";
 
 export const MemberInviteForm = ({
-	state,
-	email,
+	existingMembers,
 	isSuperAdmin,
+	addMember,
+	addUserMember,
+	closeForm,
 }: {
-	state: MemberFormState;
-	email?: string;
+	existingMembers: UsersId[];
 	isSuperAdmin?: boolean;
+	addMember: ({ userId, role }: { userId: UsersId; role: MemberRole }) => Promise<unknown>;
+	addUserMember: ({
+		email,
+		firstName,
+		lastName,
+		isSuperAdmin,
+		role,
+	}: Omit<NewUsers, "slug"> & { role: MemberRole }) => Promise<unknown>;
+	closeForm: () => void;
 }) => {
 	const community = useCommunity();
 
-	const runCreateUserWithMembership = useServerAction(actions.createUserWithMembership);
-	const runAddMember = useServerAction(actions.addMember);
-	const [isPending, startTransition] = useTransition();
-	const router = useRouter();
+	const runCreateUserWithMembership = useServerAction(addUserMember);
+	const runAddMember = useServerAction(addMember);
 
 	const form = useForm<z.infer<typeof memberInviteFormSchema>>({
 		resolver: zodResolver(memberInviteFormSchema),
 		defaultValues: {
-			email: email,
 			role: MemberRole.editor,
 			isSuperAdmin: false,
 		},
+		mode: "onChange",
 	});
+	const email = form.watch("email");
+	const emailState = form.getFieldState("email", form.formState);
+	const query = { email, limit: 1, communityId: community.id };
+	const shouldSearch = email && (!emailState.error || emailState.error.type === "alreadyMember");
+	const { data: userSuggestions, status } = client.searchUsers.useQuery({
+		queryKey: ["searchUsers", query, community.slug],
+		queryData: shouldSearch ? { query, params: { communitySlug: community.slug } } : skipToken,
+	});
+	const user = userSuggestions?.body?.[0];
+	const isPending = email && !emailState.invalid && status === "pending";
 
-	const closeForm = useCallback(
-		() => router.push(window.location.pathname!.replace(/\/add.*/, "")),
-		[]
+	const userIsAlreadyMember = useMemo(
+		() => user && existingMembers.includes(user.id),
+		[user, existingMembers]
 	);
-
-	async function onSubmit(data: z.infer<typeof memberInviteFormSchema>) {
-		if (state.state === "initial") {
-			return;
+	useEffect(() => {
+		if (userIsAlreadyMember) {
+			form.setError("email", {
+				type: "alreadyMember",
+				message: "This user is already a member",
+			});
 		}
+	}, [userIsAlreadyMember, form.setError, user, email]);
```

Comment:
For some reason the form successfully detects if a user is already a member when you enter their email the first time, but if you then edit the email and then edit it back to the user's email, the form lets you add them. I handled that on the backend, so it's not a big deal, but I'm not sure why it's happening. At first I figured I missed a dependency which is why there are these extras (I will clean that up before merging since it didn't help).

Reply:
I also didn't want to manage this error manually but I didn't find any other approach that worked. I tried validating it through the `rules` prop on the `<FormField>` component but apparently those don't run if you use a resolver. I also tried adding it as a `.refine` rule to the resolver dynamically in the useform call above (like we do in the `InputComponentConfigurationForm`) but I ended up with infinite rerenders that I couldn't figure out how to fix.